mlx/transforms.cpp

```diff
@@ -244,7 +276,8 @@ void eval(const std::vector<array>& outputs) {
     }
   };
 
-  recurse(synchronizer);
+  compute_depths(synchronizer);
+  recurse(synchronizer, false);
```

Comment:
Could you explain why `largest_branch_first` starts `false` and then switches to `true`?

Reply:
So the concept of going down the largest branch is that there are more things to detach and less likely that we 'll be keeping things around for a long time (the time it takes to process the big branch).

Obviously it is a heuristic and not a particularly good one at that. One can probably make up graphs where going down the smallest branch is the most efficient memory-wise.

Regarding the reverse direction for the synchronizer that is pretty straightforward. We know that we 'll keep all of these no matter what. As a result start from the smallest graph which will retain in memory the smallest amount of arrays to compute the second smallest graph etc. A good example is the parameter updates. If we were starting from the largest output graph, then we would keep all the gradients in memory. Otoh starting from the smallest we can throw the gradient away immediately cause we were done with the update.

So to sum up, the heuristic is as follows: Given N subgraphs where you keep only the root node of each subgraph. Start from the smallest one but process it largest branch first.

Reply:
Your explanation makes sense, I'm still pondering a bit why the strategy is a generally good one. 

Sorting the outputs by reverse depth makes sense, largest last. This is basically reverse params but more principled.

I'm wondering if there is a big difference from doing just that vs also sorting on the children?

Thinking out loud a bit:

- Largest last makes sense when there is a path between outputs for the reasons you said
- When there is no path between the outputs, presumably the order shouldn't matter much (independent sub graphs)
- I'm struggling a bit with to come up with cases where shortest last is helpful. I understand your reasoning from above, so it would imply in the graphs you were looking at we have lot's of cases where the longest path is basically fully detachable but the shortest path branch is not?

Reply:
Precisely, it depends on the large subgraph. For instance, in the following

```
 F   G    H
  \   \    \
A - B - C - D - E
```

doing shortest first keeps `H`, `G` and `F` around before computing `A`.